File: 21.merge-two-sorted-lists.py
#
# @lc app=leetcode id=21 lang=python3
#
# [21] Merge Two Sorted Lists
#
from typing import *


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next


class Solution:
    def mergeTwoLists(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        dummy = ListNode()
        cur = dummy
        cur1 = l1
        cur2 = l2
        while cur1 and cur2:
            if cur1.val <= cur2.val:
                cur.next = cur1
                cur1 = cur1.next
            else:
                cur.next = cur2
                cur2 = cur2.next
            cur = cur.next
        if cur1:
            cur.next = cur1
        if cur2:
            cur.next = cur2
        return dummy.next
# @lc code=end

File: cpp/21.merge-two-sorted-lists.cpp
/*
 * @lc app=leetcode id=21 lang=cpp
 *
 * [21] Merge Two Sorted Lists
 */
struct ListNode;
// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution
{
public:
    ListNode *mergeTwoLists(ListNode *l1, ListNode *l2)
    {
        auto res = new ListNode(), h_res = res;
        auto cur1 = l1, cur2 = l2;
        for (; cur1 != nullptr && cur2 != nullptr;)
        {
            if (cur1->val <= cur2->val)
            {
                res->next = new ListNode(cur1->val);
                res = res->next;
                cur1 = cur1->next;
            }
            else
            {
                res->next = new ListNode(cur2->val);
                res = res->next;
                cur2 = cur2->next;
            }
        }
        while (cur1 != nullptr)
        {
            res->next = new ListNode(cur1->val);
            res = res->next;
            cur1 = cur1->next;
        }

        while (cur2 != nullptr)
        {
            res->next = new ListNode(cur2->val);
            res = res->next;
            cur2 = cur2->next;
        }
        res = h_res->next;
        delete (h_res);
        return res;
    }
};
// @lc code=end
